modules/registrars/openprovider/Controllers/System/DnsController.php

```diff
@@ -45,77 +46,151 @@ public function get($params)
         $params['sld'] = $params['original']['domainObj']->getSecondLevel();
         $params['tld'] = $params['original']['domainObj']->getTopLevel();
 
-        $dnsRecordsArr = array();
         $this->domain->load(array(
             'name'      => $params['sld'],
             'extension' => $params['tld']
         ));
 
-        $dnsInfo = $this->apiHelper->getDns($this->domain);
+        try {
+            $dnsInfo = $this->apiHelper->getDns($this->domain);
+        } catch (\Exception $e) {
+            return [];
+        }
 
-        if (empty($dnsInfo)) {
+        if (empty($dnsInfo) || empty($dnsInfo['records'])) {
             return [];
         }
 
-        $supportedDnsTypes = APIConfig::$supportedDnsTypes;
-        $domainName        = $params['sld'] . '.' . $params['tld'];
+        $dnsRecords = $this->removeUnsupportedRecords($dnsInfo['records']);
 
-        foreach ($dnsInfo['records'] as $dnsRecord) {
-            if (!in_array($dnsRecord['type'], $supportedDnsTypes)) {
-                continue;
+        return $dnsRecords ?
+            $this->prepareRecordsToDisplay($dnsRecords, sprintf('%s.%s', $params['sld'], $params['tld'])) :
+            [];
+    }
+
+    /**
+     * Save the new DNS settings.
+     *
+     * @param $params
+     *
+     * @return array|string
+     */
+    public function save($params)
+    {
+        $params['sld'] = $params['original']['domainObj']->getSecondLevel();
+        $params['tld'] = $params['original']['domainObj']->getTopLevel();
+
+        $records = $this->convertToObjects($params['dnsrecords']);
+
+        $domain = $this->domain;
+        $domain->name = $params['sld'];
+        $domain->extension = $params['tld'];
+
+        try {
+            if (empty($records)) {
+                $this->apiHelper->deleteDnsRecords($domain);
+
+                return "success";
             }
 
-            $hostname = $dnsRecord['name'];
+            $dnsZone = $this->apiHelper->getDns($domain);
+            if ($dnsZone) {
+                $this->apiHelper->updateDnsRecords($domain, $records);
 
-            if ($hostname == $domainName) {
-                $hostname = '';
-            } else {
-                $pos = stripos($hostname, '.' . $domainName);
-                if ($pos !== false) {
-                    $hostname = substr($hostname, 0, $pos);
-                }
+                return "success";
             }
 
-            $prio            = is_numeric($dnsRecord['prio']) ? $dnsRecord['prio'] : '';
-            $dnsRecordsArr[] = [
-                'hostname' => $hostname,
-                'type'     => $dnsRecord['type'],
-                'address'  => $dnsRecord['value'],
-                'priority' => $prio
+            $this->apiHelper->createDnsRecords($domain, $records);
+        } catch (\Exception $e) {
+            return [
+                'error' => $e->getMessage(),
             ];
         }
 
-        return $dnsRecordsArr;
+        return "success";
```

Comment:
It will be better better to delete it from here and add it after
`$this->apiHelper->createDnsRecords($domain, $records);`
to unify your code